### ISSUE TITLE
Sync current token may be ahead of event stream cache position using workers

### ISSUE THREAD

**Issue by Fizzadar** [matrix-org/synapse#14158]
Based on my initial work investigating sync cache races in: https://github.com/matrix-org/synapse/issues/14154

Sync makes use of the event stream cache to determine whether a room has changed between the since & current tokens. This is then used to limit the number of rooms events are queried for in `get_room_events_stream_for_rooms`. After discovering the cache invalidation races above I added a quick log line for this: https://github.com/beeper/synapse/commit/62497db551651e2ebfb9241a3521a9410aa98706 (after https://github.com/beeper/synapse/commit/529715536c3d0783ba76093e6826c5516f934bb0).

And it logs! Only a very small handful of occurrences over the last ~5 days and the position difference has so far been 1 every time. I suspect this may also occur against other stream caches but have not confirmed.

The worry here is if an event was sent within the gap it may be missed from an incremental sync which is especially bad because the user will never see or know about this event unless they re-init sync (or the client backfills it?).

One solution to this is to implement a waiting mechanism on `StreamCache` so a worker can wait for the cache to catch up with the current token for a given sync before fetching data. Because this is super rare and even when it happens it's a tiny position difference this would probably have negligable impact in sync performance and provide a shield against cache invalidation races over replication.

**Comment by Fizzadar** [comment]
Confirmed via the log change above, this is actually not that rare (😱). We have large bursts of these in sync workers:

```
Checking entity in DeviceInboxStreamChangeCache stream cache cache with
    current position 42696041 ahead of  the max stream cache position 42696040
```

There's even more in federation sender instances `DeviceFederationOutboxStreamChangeCache`, pretty much a constant flow of these:

```
Checking entity in DeviceFederationOutboxStreamChangeCache stream cache cache with
    current position 42696140 ahead of the max stream cache position 4269609
```

Finally the background work has a constant (1/s) flow of such errors on the `_curr_state_delta_stream_cache` cache, where in that case the cache often appears to be lagging significantly:

```
Checking entity in _curr_state_delta_stream_cache stream cache cache with
    current position 540123283 ahead of the max stream cache position 540123279
```

**Comment by Fizzadar** [comment]
Implemented wait-for fix option: https://github.com/matrix-org/synapse/pull/14269

**Comment by richvdh** [comment]
err, interesting. That doesn't sound good.

> There's even more in federation sender instances `DeviceFederationOutboxStreamChangeCache`, pretty much a constant flow of these:
> 
> ```
> Checking entity in DeviceFederationOutboxStreamChangeCache stream cache cache with
>     current position 42696140 ahead of the max stream cache position 42696092
> ```

This sounds very much like it could be the cause of https://github.com/matrix-org/synapse/issues/14251 and other device-list-tracking bugs we've been tearing our hair out over.

For those of us struggling to keep up at the back: are you able to explain the mechanics of this race? How does the "current position" come to be ahead of the "max stream cache position"?

I wasn't able to find where exactly you introduced these log lines - can you point me to a relevant diff?

**Comment by erikjohnston** [comment]
> At that point, `get_current_token` will immediately point at that token

`get_current_token` should only return the new token when we've finished persisting that token (and it gets more complicated with multi-writer). So, as long as we poke the stream change cache at the same time as calling `get_next` (or `advance` in replication case) this should be fine.

I think part of the confusion here is that there is not a one-to-one mapping between stream caches and and stream ID generators. For example, the device inbox ID generator feeds into *both* `DeviceInboxStreamChangeCache` and `DeviceFederationOutboxStreamChangeCache`:

https://github.com/matrix-org/synapse/blob/8c94dd3a277d4e11192f98a9ca32cb6638606b66/synapse/storage/databases/main/deviceinbox.py#L688-L698

So it's not unexpected if one *xor* the other stream change cache lags behind the current token.

---

Now, that's not to say there isn't a problem here. I'd be tempted for some of these stream caches to pass in the relevant stream ID generator, and then log if we are inserting a change with a lower stream ID than returned by `get_current_token()` (and maybe also `get_current_token(instance_name)` for multi writer streams). This may have some a false positive in process replication rows where we advance the ID gen and *then* update the stream change cache, e.g.:

https://github.com/matrix-org/synapse/blob/8c94dd3a277d4e11192f98a9ca32cb6638606b66/synapse/storage/databases/main/deviceinbox.py#L141-L146

**Comment by Fizzadar** [comment]
I think I've been misunderstanding `get_current_token` - I was thinking it fetched it from the database but it's not it's the "known latest position" which is effectively the same as the waiting thing. Which makes this more confusing - how are we getting in a situation where the event ID generator has a token ahead of the events stream cache?

---

Re: device inbox I discovered this also where the two stream caches share one ID generator. My WIP PR https://github.com/matrix-org/synapse/pull/14286 addresses this by making sure both stream caches are always updated (to enable the waiting logic).

**Comment by Fizzadar** [comment]
Digging into this a bit, the only place the stream ID is advanced is:

https://github.com/matrix-org/synapse/blob/2d0ba3f89aaf9545d81c4027500e543ec70b68a6/synapse/storage/databases/main/events_worker.py#L285-L295

And the change should be propagated at the same time here:

https://github.com/matrix-org/synapse/blob/2d0ba3f89aaf9545d81c4027500e543ec70b68a6/synapse/storage/databases/main/cache.py#L185-L210

Is it possible that we see a `EventsStreamCurrentStateRow.TypeId` row in a separate batch to `EventsStreamEventRow.TypeId`? Both of those would advance the token but only the first would flag the change in the stream cache at that token.

**Comment by Fizzadar** [comment]
Should the `entity_has_changed` call just be moved into the `EventsWorkerStore.process_replication_rows`? It seems to me the token and change cache are deeply interlinked and keeping the code together would make sense (like `DeviceInboxWorkerStore` does).

**Comment by Fizzadar** [comment]
I’m keen to work on this plan and should have some time over the next few weeks, feel free to assign to me :) There’s some precursor work I have identified as well involving cleanup of the slaved stores so will start there.

**Comment by Fizzadar** [comment]
Small update - the new `ReceiptsRoomChangeCache` also triggers the same issue, I threw up a very quick debug change to our sync workers, looks like they are getting calls to `process_replication_data` with an empty row list (https://github.com/beeper/synapse/pull/35/commits/2610e255f4a19362f4008a9ab1aa71010dfc4224).

This is expected for `POSITION` commands... but as far as I can tell these are not sent for receipts, so how is there ever an empty list here!

**Comment by Fizzadar** [comment]
I think I may have figured out why this happens with the `EventsRoomStreamChangeCache`! 

- Sync workers call `get_unread_event_push_actions_by_room_for_user` for rooms
- This executes check on the cache _inside_ a transaction, and thus inside a thread
- Python unlocks the GIL every Xms for threads, so it's possible this thread is switched to mid-way through the `process_replication_rows` calls
- Because of class resolution order the `EventsWorkerStore.process_replication_rows` is called before `CacheInvalidationWorkerStore.process_replication_rows` which opens up this window

This explains why we're not seeing this particular log causing major issues and why it's so rare; I have pushed some logging to our fork to confirm the above is the case. Fixing this is relatively easy by making sure the stream cache is updated before the ID generator.

**Comment by erikjohnston** [comment]
> Fixing this is relatively easy by making sure the stream cache is updated before the ID generator

Should we add an assertion in the stream change cache where we assert that when we mark something as changed the stream ID is strictly greater than the current token?

**Comment by erikjohnston** [comment]
Also, :scream:

**Comment by Fizzadar** [comment]
> Should we add an assertion in the stream change cache where we assert that when we mark something as changed the stream ID is strictly greater than the current token?

Yes! That’d be ideal, will probably require passing the relevant id generator into the stream change cache but that seems sensible anyway…

**Comment by Fizzadar** [comment]
I have added that to my PR, enabled only for the events stream, in https://github.com/matrix-org/synapse/pull/14648/commits/5ee04253c1e2d33e3bad9d8c74422cf468b304b0. I think enabling for other streams is best left to a new PR?

**Comment by erikjohnston** [comment]
@Fizzadar which PR?

**Comment by clokep** [comment]
> @Fizzadar which PR?

Seem to be #14648

**Comment by richvdh** [comment]
As I understand it, this has been partially addressed by #14723? Is anyone able to provide an update on what remains to be done here?

**Comment by Fizzadar** [comment]
> As I understand it, this has been partially addressed by #14723? Is anyone able to provide an update on what remains to be done here?

This should indeed resolve the issue once and for all - I'm waiting on rolling 1.75  to our deployment to confirm (as we have the logs indicating the issue) but I'm pretty sure that PR solves it. Will update here/close issue once that's live.

**Comment by Fizzadar** [comment]
So this is still happening on our deployment but it did occur to me this morning - could this be a side effect of multiple event persisters, where one is running behind the others?

If I'm following the [`ReplicationCommandHandler._process_position`](https://github.com/matrix-org/synapse/blob/fc35e0673f5b46ea0f5e53ef15626b14a452ca82/synapse/replication/tcp/handler.py#L557) code properly it looks like replication positions may be processed out of order. This would then mean changes are read/written to the change caches out of order from time to time?

Note: this probably means the assertion that changes are incremental only on cache update would blow up in these situations...

Does it make sense to force and order here and prevent out-of-order event persisters? Would mean any slow persister holds back sync/replication everywhere but they really shouldn't be that far behind each other...
